runtime/major_gc.c

```diff
@@ -114,30 +129,150 @@ static unsigned long major_gc_counter = 0;
+
+  for( entry = 0; entry < mark_stack_count ; entry++ ) {
+    mark_entry me = mark_stack[entry];
+    value* block_op = Op_val(me.block);
+    uintnat chunk_addr = 0, chunk_addr_below = 0;
+
+    caml_skiplist_find_below(&chunk_sklist, (uintnat)me.block, &chunk_addr,
+                              &chunk_addr_below);
+
+    if( chunk_addr != 0 && me.block >= chunk_addr
+        && me.block < chunk_addr_below ) {
+
+      if( Chunk_redarken_start(chunk_addr) > block_op ) {
+        Chunk_redarken_start(chunk_addr) = block_op;
+      }
+
+      if( Chunk_redarken_end(chunk_addr) < block_op ) {
+        Chunk_redarken_end(chunk_addr) = block_op;
+      }
+
+      Chunk_requires_redarken(chunk_addr) = 1;
+
+      if( redarken_first_chunk == NULL
+          || redarken_first_chunk > (char*)chunk_addr ) {
+        redarken_first_chunk = (char*)chunk_addr;
+      }
+    }
+  }
+
+  caml_skiplist_empty(&chunk_sklist);
+
+  caml_gc_message(0x08, "Mark stack overflow.\n");
+
+  stk->count = 0;
+}
+
+static void realloc_mark_stack (struct mark_stack* stk)
 {
-  value *new;
+  mark_entry* new;
+  uintnat mark_stack_bsize = stk->size * sizeof(mark_entry);
 
-  CAMLassert (gray_vals_cur == gray_vals_end);
-  if (1 /* FIXME: handle overflow if
-           gray_vals_size < caml_stat_heap_wsz / 32*/){
-    caml_gc_message (0x08, "Growing gray_vals to %"
+  if ( mark_stack_bsize < Caml_state->stat_heap_wsz / 32
+      || mark_stack_bsize < MARK_STACK_INIT_SIZE*sizeof(mark_entry) ) {
+    caml_gc_message (0x08, "Growing mark stack to %"
                            ARCH_INTNAT_PRINTF_FORMAT "uk bytes\n",
-                     (intnat) gray_vals_size * sizeof (value) / 512);
-    new = (value *) caml_stat_resize_noexc ((char *) gray_vals,
-                                            2 * gray_vals_size *
-                                            sizeof (value));
-    if (new == NULL){
-      caml_fatal_error("No room for growing gray_vals");
-    }else{
-      gray_vals = new;
-      gray_vals_cur = gray_vals + gray_vals_size;
-      gray_vals_size *= 2;
-      gray_vals_end = gray_vals + gray_vals_size;
+                     (intnat) mark_stack_bsize * 2 / 1024);
+
+    new = (mark_entry*) caml_stat_resize_noexc ((char*) stk->stack,
+                                                2 * mark_stack_bsize);
+    if (new != NULL) {
+      stk->stack = new;
+      stk->size *= 2;
+      return;
     }
   }
+
+  caml_gc_message (0x08, "No room for growing mark stack. Pruning..\n");
+  mark_stack_prune(stk);
+}
+
+/* This function pushes the provided mark_entry [me] on to the current mark
+   stk [stk]. It first checks, if the block is small enough, whether there
+   are any fields we would actually do mark work on. If so then it enqueues
+   the entry. */
+Caml_inline void mark_stack_push(struct mark_stack* stk, mark_entry me,
+                                  intnat* work)
+{
+  value v;
+  int i, block_end = Wosize_val(me.block), end;
```

Comment:
```suggestion
  int i, block_wsz = Wosize_val(me.block), end;
```

The variable name `block_end` sounds like it is a pointer to the end of the block.

Reply:
Yes, that's clearer.